[JENKINS-49971] Fix a race condition in Util.loadFile()

dtrebbien opened the pull request:
If the file is deleted in between when its existence is checked and the
file is opened for reading, then the method will fail to return an empty
string.

### Proposed changelog entries

* Prevent FileNotFoundException in `hudson.Util.loadFile()` in the case of race condition with file deletion

### Submitter checklist

- [ ] JIRA issue is well described
- [ ] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [ ] Appropriate autotests or explanation to why this change has no tests
- [ ] For dependency updates: links to external changelogs and, if possible, full diffs

<!-- For new API and extension points: Link to the reference implementation in open-source (or example in Javadoc) -->

### Desired reviewers

@oleg-nenashev

daniel-beck commented on core/src/main/java/hudson/Util.java:
```diff
@@ -179,28 +179,57 @@ public static String replaceMacro(@CheckForNull String s, @Nonnull VariableResol
+        try {
+            bytes = Files.readAllBytes(path);
+        } catch (NoSuchFileException ignored) {
+            return "";
```
It is unclear to me why this is preferable over just wrapping it in an `IOException` that already gets thrown.

dtrebbien replied:
The existing code initialized the capacity of the StringBuilder to `(int)logfile.length()` outside the try-with-resources statement that opened the file.

It is possible to move the initialization of the StringBuilder within the try-with-resources statement with something like:
```java
@Nonnull
public static String loadFile(@Nonnull File logfile, @Nonnull Charset charset) throws IOException {
    StringBuilder str;

    try (BufferedReader r = Files.newBufferedReader(fileToPath(logfile), charset)) {
        str = new StringBuilder((int)logfile.length());

        char[] buf = new char[1024];
        int len;
        while ((len = r.read(buf, 0, buf.length)) > 0)
            str.append(buf, 0, len);
    } catch (NoSuchFileException e) {
        return "";
    }

    return str.toString();
}
```

However, I looked at the Oracle JDK implementation of readAllBytes() and I thought that it was better at handling special cases such as avoiding casting a too-large `long` value to `int`, the file being truncated while it is read, or the file being written to as it is being read:

```java
/**
 * The maximum size of array to allocate.
 * Some VMs reserve some header words in an array.
 * Attempts to allocate larger arrays may result in
 * OutOfMemoryError: Requested array size exceeds VM limit
 */
private static final int MAX_BUFFER_SIZE = Integer.MAX_VALUE - 8;

private static byte[] read(InputStream source, int initialSize) throws IOException {
    int capacity = initialSize;
    byte[] buf = new byte[capacity];
    int nread = 0;
    int n;
    for (;;) {
        // read to EOF which may read more or less than initialSize (eg: file
        // is truncated while we are reading)
        while ((n = source.read(buf, nread, capacity - nread)) > 0)
            nread += n;

        // if last call to source.read() returned -1, we are done
        // otherwise, try to read one more byte; if that failed we're done too
        if (n < 0 || (n = source.read()) < 0)
            break;

        // one more byte was read; need to allocate a larger buffer
        if (capacity <= MAX_BUFFER_SIZE - capacity) {
            capacity = Math.max(capacity << 1, BUFFER_SIZE);
        } else {
            if (capacity == MAX_BUFFER_SIZE)
                throw new OutOfMemoryError("Required array size too large");
            capacity = MAX_BUFFER_SIZE;
        }
        buf = Arrays.copyOf(buf, capacity);
        buf[nread++] = (byte)n;
    }
    return (capacity == nread) ? buf : Arrays.copyOf(buf, nread);
}

public static byte[] readAllBytes(Path path) throws IOException {
    try (SeekableByteChannel sbc = Files.newByteChannel(path);
         InputStream in = Channels.newInputStream(sbc)) {
        long size = sbc.size();
        if (size > (long)MAX_BUFFER_SIZE)
            throw new OutOfMemoryError("Required array size too large");

        return read(in, (int)size);
    }
}
```

I also liked the use of SeekableByteChannel.size() instead of File.length().

Unfortunately, there isn't a more specialized standard API for reading all characters in a file, say Files.readAllChars(Path, Charset), so I have resorted to calling Files.readAllBytes() to read all bytes, and then constructing the String from the bytes and specified charset.

I don't know this project's philosophy on whether it is better to reuse existing code or replicate and specialize code for each particular case. If the latter, I could add another commit that would do what a hypothetical Files.readAllChars(Path, Charset) API would do.

oleg-nenashev commented:
I will try to review it next week, sorry for the delay.
The merge conflict needs to be resolved tho

dtrebbien committed:
Fix a race condition in Util.loadFile()

If the file is deleted in between when its existence is checked and the
file is opened for reading, then the method will fail to return an empty
string.

dtrebbien commented:
Rebased to resolve the conflict.

oleg-nenashev reviewed (changes requested):
Why not just https://commons.apache.org/proper/commons-io/javadocs/api-2.5/org/apache/commons/io/FileUtils.html#readFileToString(java.io.File,%20java.nio.charset.Charset) with FileNotFound exception handling? 

Such approach is not compatible with virtual Filesystem introduced by `Path`, but IIUC we do not need that for this utility method. At least for now...

dtrebbien committed:
Switch to using FileUtils.readFileToString()

dtrebbien commented:
FileUtils.readFileToString() looks good to me. I have pushed a new commit that switches to using this API.

dtrebbien commented:
Merge conflict resolved.

oleg-nenashev commented:
@dtrebbien It may deserve backporting. Could you please create a JIRA ticket? My plan is to merge it to the next weekly

dtrebbien commented:
Sure: https://issues.jenkins-ci.org/browse/JENKINS-49971
